tools/ot/src/main/java/dk/alexandra/fresco/tools/ot/base/NaorPinkasOt.java

```diff
@@ -0,0 +1,323 @@
+   */
+  private StrictBitVector recoverTrueMessage(byte[] encryptedZeroMessage,
+      byte[] encryptedOneMessage, byte[] seed, boolean choiceBit) {
+    if (encryptedZeroMessage.length != encryptedOneMessage.length) {
+      throw new MaliciousException(
+          "The length of the two choice messages is not equal");
+    }
+    byte[] unpaddedMessage;
+    if (choiceBit == false) {
+      unpaddedMessage = unpadMessage(encryptedZeroMessage, seed);
+    } else {
+      unpaddedMessage = unpadMessage(encryptedOneMessage, seed);
+    }
+    return new StrictBitVector(unpaddedMessage);
+  }
+
+  /**
+   * Completes the sender's part of the Naor-Pinkas OT.
+   *
+   * @return The two random messages sent by the sender.
+   */
+  private Pair<byte[], byte[]> sendBytesOt() {
+    // Pick random element c
+    BigInteger c = randNum.nextBigInteger(params.getP());
+    network.send(otherId, c.toByteArray());
+    BigInteger publicKeyZero = new BigInteger(network.receive(otherId));
+    // publicKeyOne = c / keyZero
+    BigInteger publicKeyOne = publicKeyZero.modInverse(params.getP())
+        .multiply(c);
+    byte[] messageZero = new byte[hashDigest.getDigestLength()];
+    byte[] messageOne = new byte[hashDigest.getDigestLength()];
+    BigInteger encZero = encryptMessage(publicKeyZero, messageZero);
+    BigInteger encOne = encryptMessage(publicKeyOne, messageOne);
+    network.send(otherId, encZero.toByteArray());
+    network.send(otherId, encOne.toByteArray());
+    return new Pair<>(messageZero, messageOne);
+  }
+
+  /**
+   * Completes the receiver's part of the Naor-Pinkas OT.
+   *
+   * @return The message received
+   */
+  private byte[] receiveByteOt(Boolean choiceBit) {
+    BigInteger c = new BigInteger(network.receive(otherId));
+    // Pick random element privateKey
+    BigInteger privateKey = randNum.nextBigInteger(params.getP());
+    // publicKeySigma = G^privateKey mod P
+    BigInteger publicKeySigma = params.getG().modPow(privateKey, params.getP());
+    // publicKeyNotSigma = c / publicKeySigma mod P
+    BigInteger publicKeyNotSigma = publicKeySigma.modInverse(params.getP())
+        .multiply(c);
+    if (choiceBit == false) {
+      network.send(otherId, publicKeySigma.toByteArray());
+    } else {
+      network.send(otherId, publicKeyNotSigma.toByteArray());
+    }
+    BigInteger encZero = new BigInteger(network.receive(otherId));
+    BigInteger encOne = new BigInteger(network.receive(otherId));
+    byte[] message;
+    if (choiceBit == false) {
+      message = decryptMessage(encZero, privateKey);
+    } else {
+      message = decryptMessage(encOne, privateKey);
+    }
+    return message;
+  }
+
+  /**
+   * Completes the internal Naor-Pinkas encryption.
+   *
+   * @param publicKey
+   *          The public key to encrypt with
+   * @param message
+   *          The message to encrypt
+   * @return The ciphertext
+   */
+  private BigInteger encryptMessage(BigInteger publicKey,
```

Comment:
Does this actually encrypt the message? As I understand Naor Pinkas the ciphertext should be _(g<sup>r</sup>, H(PK<sup>r</sup>) &oplus; M)_. This seems to only return the _g<sup>r</sup>_ part and copy the _H(PK<sup>r</sup>)_ part into the `message` array. Am I missing something?

Reply:
Ok, I understand now. You do the actual OTP encryption in the `padMessage` method. In that case, there is just an issue with the naming and/or documentation of this method. I.e., the method does not really encrypt but does computes some data used to form the ciphertext. Also, the message is not actually encrypted but used to return part this data as a side-effect.

I think we should:
- Rename the method and update the javadoc
- Return both values either as a Pair or as some appropriate inner class representing the entire output of this method
- Consinder combining this method and the `padMessage` method into a single method doing the full encryption.

Reply:
I use the notation of the original paper, where these are called encryptions. I don't want to combine things into an even bigger method because this is something @pffrandsen is firmly against (and I completely agree). 
I am not sure what you want me represent as a pair? Ciphertext and key?
I can rename things though. How about PrepareOTP?

Reply:
I have changed things a bit according to @GuutBoy 's suggestions in commit 06f47313